aten/src/ATen/cuda/detail/DeviceThreadHandles.h

```diff
@@ -0,0 +1,147 @@
+    std::unordered_map<int, std::vector<Handle>> created_handles;
+    std::unordered_map<int, std::vector<Handle_t>> available_handles;
+
+    // PoolWindow lazily creates and caches the handles that a particular thread is using,
+    // so in the common case handle access doesn't incur either handle creation or a mutex lock.
+    class PoolWindow
+    {
+    DeviceThreadHandlePool &parent;
+    public:
+    PoolWindow(DeviceThreadHandlePool &parent): parent(parent) {}
+    ~PoolWindow(){ release(); }
+
+    Handle_t reserve(int device)
+    {
+        // If this thread already has a handle for this device, return it
+        if(my_handles.find(device) != my_handles.end())
+        return my_handles[device];
+
+        // otherwise, either grab a handle from the pool if one is available,
+        // or if not, create a new one.
+        std::lock_guard<std::mutex> guard(parent.mutex);
+
+        if(parent.available_handles[device].size() > 0)
+        {
+        my_handles[device] = parent.available_handles[device].back();
+        parent.available_handles[device].pop_back();
+        }
+        else
+        {
+        // In local testing, I do observe that emplace_back sometimes routes through temporaries
+        // that incur move-constructor and destructor calls.  See comments in Handle above.
+        parent.created_handles[device].emplace_back(true /*create*/);
+        my_handles[device] = parent.created_handles[device].back().handle;
+        }
+
+        return my_handles[device];
+    }
+
+    private:
+    // Stores the per-device handles currently owned by this thread
+    std::unordered_map<int, Handle_t> my_handles;
+
+    // Called by the destructor.  Releases this thread's handles back into the pool.
+    void release()
+    {
+        // The conditional check below is certainly worthwhile for efficiency.
+        // However, it also serves another purpose:  Without the conditional,
+        // as of cuda V9.0.176 and torch.backends.cudnn.version() = 7005,
+        // we observe weird nondeterministic hangs on Windows when the process first
+        // attempts to create a cudnn handle.  Example with added debug print statements:
+        // https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-test2/19238/console
+        // The print statements reveal that when these hangs occur, the thread that is attempting
+        // to create a cudnn handle for the first time hangs on the call to cudnnCreate itself.
+        // These hangs have never manifested on anything but that particular Windows build.
+        // All other builds seem fine.
+        if(my_handles.size() > 0)
+        {
+        std::lock_guard<std::mutex> guard(parent.mutex);
+        for(auto d_h : my_handles)
+            parent.available_handles[d_h.first].push_back(d_h.second);
+        }
+    }
+    };
+
+    PoolWindow *newPoolWindow() {
```

Comment:
Add a comment that this function must be thread safe, because it's called without a mutex to create the thread_local unique_ptrs.